src/MSBuild/XMake.cs

```diff
@@ -2514,87 +2515,195 @@ private static bool GatherAutoResponseFileSwitchesFromFullPath(string autoRespon
             return invokeBuild;
         }
 
-        private static bool ProcessTerminalLoggerConfiguration(CommandLineSwitches commandLineSwitches)
+        private static bool ProcessTerminalLoggerConfiguration(CommandLineSwitches commandLineSwitches, out string aggregatedParameters)
         {
-            string terminalloggerArg;
+            aggregatedParameters = AggregateParameters(commandLineSwitches);
+            string defaultValue = FindDefaultValue(aggregatedParameters);
 
-            // Command line wins, so check it first
-            if (commandLineSwitches.IsParameterizedSwitchSet(CommandLineSwitches.ParameterizedSwitch.TerminalLogger))
+            string terminalLoggerArg = null;
+            if (!TryFromCommandLine(commandLineSwitches) && !TryFromEnvironmentVariables())
             {
+                ApplyDefault();
+            }
+
+            terminalLoggerArg = NormalizeIntoBooleanValues();
+
+            bool useTerminalLogger = false;
+            if (!TrueOrFalse())
+            {
+                ItMustBeAuto();
+            }
+
+            return KnownTelemetry.LoggingConfigurationTelemetry.TerminalLogger = useTerminalLogger;
+
+            static bool DoesEnvironmentSupportTerminalLogger()
+            {
+                (var acceptAnsiColorCodes, var outputIsScreen, s_originalConsoleMode) = NativeMethodsShared.QueryIsScreenAndTryEnableAnsiColorCodes();
+
+                if (!outputIsScreen)
+                {
+                    s_globalMessagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage(ResourceUtilities.GetResourceString("TerminalLoggerNotUsedRedirected"), MessageImportance.Low));
+                    return false;
+                }
+
+                // TerminalLogger is not used if the terminal does not support ANSI/VT100 escape sequences.
+                if (!acceptAnsiColorCodes)
+                {
+                    s_globalMessagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage(ResourceUtilities.GetResourceString("TerminalLoggerNotUsedNotSupported"), MessageImportance.Low));
+                    return false;
+                }
+
+                if (Traits.Instance.EscapeHatches.EnsureStdOutForChildNodesIsPrimaryStdout)
+                {
+                    s_globalMessagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage(ResourceUtilities.GetResourceString("TerminalLoggerNotUsedDisabled"), MessageImportance.Low));
+                    return false;
+                }
+
+                return true;
+            }
+
+            string FindDefaultValue(string s)
+            {
+                // Find default configuration so it is part of telemetry even when default is not used.
+                // Default can be stored in /tlp:default=true|false|on|off|auto
```

Comment:
Question: I can see defaulting to "auto", and I can see defaulting to "off" (after we ship, if you like old-n-busted better). But is there a case to default to "on"? It feels like that's dangerous since it'll junk up logs in a bunch of situations.

(In retrospect it doesn't make sense for the current env vars to support `true` either . . .)

Reply:
This is good point. When I though about I personally prefer to give customer freedom to decide, playing with undocumented /tlp is "for experts only" and they should know better than enable it on CI.
One possible usage is: `msbuild /tl:true > t.log; cat t.log`.  When you `cat t.log` it will replay log on terminal with all escape chars and such. One can also investigate `raw` form of output. I used it few time to investigate and troubleshoot, but outside of this niche use cases, it might be risky. 
I consider `/tl:true` more risky than `/tlp:default=true` in this context as people can do that by mistake as `true` could be instinctive 'nobody-has-time-for-read-manual' kind of value they choose.
Maybe we can shall `true | on` by `force`. 
@rainersigwald @baronfel what do you think about ^?

Reply:
there's not really a use case for `on/true` here, but IMO it's more consistent to describe this parameter in terms of the valid values for the MSBUILDTERMINALLOGGER environment variable, since it's answering the question "what default for `this_other_value` should I use if non is explicitly provided".